commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/RecordDefinitionTranslator.java

```diff
@@ -0,0 +1,733 @@
+                    typeID = type_string;
+                }
+                else if (srcType == HpccSrcType.UTF16LE || srcType == HpccSrcType.UTF16BE)
+                {
+                    typeID = type_unicode;
+                }
+                else if (srcType == HpccSrcType.UTF8)
+                {
+                    typeID = type_utf8;
+                }
+                else if (srcType == HpccSrcType.QSTRING)
+                {
+                    typeID = type_qstring;
+                }
+                else
+                {
+                    throw new Exception(
+                            "Unable to convert type to json. Encountered unexpected string source type: " + srcType);
+                }
+
+                if (field.isFixed() == false)
+                {
+                    typeID |= FLAG_UNKNOWN_SIZE;
+                }
+
+                return typeID;
+            }
+            case VAR_STRING:
+            {
+                int typeID = -1;
+                HpccSrcType srcType = field.getSourceType();
+                if (srcType == HpccSrcType.SINGLE_BYTE_CHAR)
+                {
+                    typeID = type_varstring;
+                }
+                else if (srcType == HpccSrcType.UTF16LE || srcType == HpccSrcType.UTF16BE)
+                {
+                    typeID = type_varunicode;
+                }
+                else
+                {
+                    throw new Exception(
+                            "Unable to convert type to json. Encountered unexpected varstring source type: " + srcType);
+                }
+
+                if (field.isFixed() == false)
+                {
+                    typeID |= FLAG_UNKNOWN_SIZE;
+                }
+
+                return typeID;
+            }
+            case RECORD:
+            {
+                int typeID = type_record;
+                if (field.isFixed() == false)
+                {
+                    typeID |= FLAG_UNKNOWN_SIZE;
+                }
+                return typeID;
+            }
+            default:
+            {
+                throw new Exception("Unable to generate JSON for field : " + field.getFieldName()
+                        + " with unknown type: " + field.getFieldType().description());
+            }
+        }
+    }
+
+    private static int getTypeHash(FieldDef field) throws Exception
+    {
+        int numHashComponents = 2 + field.getNumDefs();
+        if (field.getFieldType() == FieldType.DECIMAL)
+        {
+            numHashComponents += 2;
+        }
+
+        long[] hashComponents = new long[numHashComponents];
+        hashComponents[0] = getTypeID(field);
+        hashComponents[1] = field.getDataLen();
+
+        int hashCompIndex = 2;
+        for (int i = 0; i < field.getNumDefs(); i++, hashCompIndex++)
+        {
+            hashComponents[hashCompIndex] = getTypeHash(field.getDef(i));
+        }
+
+        if (field.getFieldType() == FieldType.DECIMAL)
```

Comment:
will DECIMALS ever have field.getNumDefs() > 2?
if not, it seems this block could move up to line 512 for  readability